### ISSUE TITLE
run_tests.py scan fails on fresh installation

### ISSUE THREAD

**Issue by josiahjohnston** [switch-model/switch#7]
Reed had trouble with run_tests.py on a fresh install. I'll ask him to assign himself as well to join in the conversation.

```bash
Reeds-MacBook-Air-2:/ reedHaubenstock$ ls -l /Users/reedHaubenstock/Desktop/switch_py
/Users/reedHaubenstock/Desktop/switch_py:
total 72
-rwxr-xr-x@  1 reedHaubenstock  staff    342 Jul  8 20:15 AUTHORS
-rwxr-xr-x@  1 reedHaubenstock  staff   1104 Jul  8 20:15 INSTALL
-rwxr-xr-x@  1 reedHaubenstock  staff  11555 Jul  8 20:15 LICENSE
-rwxr-xr-x@  1 reedHaubenstock  staff    732 Jul  8 20:15 LICENSE.BOILERPLATE
-rwxr-xr-x@  1 reedHaubenstock  staff    956 Jul  8 20:15 README
drwxr-xr-x@ 30 reedHaubenstock  staff   1020 Jul  8 20:15 doc
drwxr-xr-x@  9 reedHaubenstock  staff    306 Jul 14 12:24 examples
-rwxr-xr-x@  1 reedHaubenstock  staff   2257 Jul  8 20:15 run_tests.py
-rw-r--r--+  1 reedHaubenstock  staff   1609 Jul 14 12:19 run_tests.pyc
drwxr-xr-x@ 15 reedHaubenstock  staff    510 Jul  8 20:15 sandbox_dev
drwxr-xr-x@ 32 reedHaubenstock  staff   1088 Jul 14 12:18 switch_mod
drwxr-xr-x@  4 reedHaubenstock  staff    136 Jul  8 20:15 tests
Reeds-MacBook-Air-2:/ reedHaubenstock$ echo $PYTHONPATH
:/Users/reedHaubenstock/Desktop/switch_py
Reeds-MacBook-Air-2:/ reedHaubenstock$ cd Users/reedHaubenstock/Desktop/switch_py 
Reeds-MacBook-Air-2:switch_py reedHaubenstock$ ./run_tests.py

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
```

**Comment by josiahjohnston** [comment]
Reed, what do you get if you try running a doctest and unittest directly with these commands? 
```bash
# Execute a doctest of timescales
cd /Users/reedHaubenstock/Desktop/switch_py/switch_mod
python -m doctest -v timescales.py
# Execute a unit test
cd /Users/reedHaubenstock/Desktop/switch_py/tests
python utilities_test.py
```

**Comment by mseaborn** [comment]
What does "python --version" print on the system that has this problem?  I'm not sure how likely it is that you're using a version before 2.7, but that might cause this.

What version of Mac OS X are you using?

**Comment by rhaubenstock** [comment]
The command "python --version" prints Python 2.7.3 when I run it in terminal. I am using Mac OS X Yosemite Version 10.10.1. However this problem only occurs when I run the command "./run_tests.py" while in the switch_py project folder. When I run the command "python -m doctest -v timescales.py" 
in the switch_mod directory I get the following output: 

"Trying:
    from switch_mod.utilities import define_AbstractModel
Expecting nothing
**********************************************************************
File "timescales.py", line 8, in timescales
Failed example:
    from switch_mod.utilities import define_AbstractModel
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest timescales[0]>", line 1, in <module>
        from switch_mod.utilities import define_AbstractModel
    ImportError: No module named switch_mod.utilities
Trying:
    model = define_AbstractModel('timescales')
Expecting nothing
**********************************************************************
File "timescales.py", line 9, in timescales
Failed example:
    model = define_AbstractModel('timescales')
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest timescales[1]>", line 1, in <module>
        model = define_AbstractModel('timescales')
    NameError: name 'define_AbstractModel' is not defined
Trying:
    instance = model.load_inputs(inputs_dir='test_dat')
Expecting nothing
**********************************************************************
File "timescales.py", line 10, in timescales
Failed example:
    instance = model.load_inputs(inputs_dir='test_dat')
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest timescales[2]>", line 1, in <module>
        instance = model.load_inputs(inputs_dir='test_dat')
    NameError: name 'model' is not defined
Trying:
    from switch_mod.utilities import define_AbstractModel
Expecting nothing
**********************************************************************
File "timescales.py", line 223, in timescales.define_components
Failed example:
    from switch_mod.utilities import define_AbstractModel
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest timescales.define_components[0]>", line 1, in <module>
        from switch_mod.utilities import define_AbstractModel
    ImportError: No module named switch_mod.utilities
Trying:
    model = define_AbstractModel('timescales')
Expecting nothing
**********************************************************************
File "timescales.py", line 224, in timescales.define_components
Failed example:
    model = define_AbstractModel('timescales')
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest timescales.define_components[1]>", line 1, in <module>
        model = define_AbstractModel('timescales')
    NameError: name 'define_AbstractModel' is not defined
Trying:
    instance = model.create('test_dat/timescales.dat')
Expecting nothing
**********************************************************************
File "timescales.py", line 225, in timescales.define_components
Failed example:
    instance = model.create('test_dat/timescales.dat')
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest timescales.define_components[2]>", line 1, in <module>
        instance = model.create('test_dat/timescales.dat')
    NameError: name 'model' is not defined
Trying:
    instance = model.create('test_dat/timescales_bad_weights.dat')
Expecting:
    Traceback (most recent call last):
        ...
    ValueError: BuildCheck 'validate_time_weights' identified error with index '2020'
**********************************************************************
File "timescales.py", line 226, in timescales.define_components
Failed example:
    instance = model.create('test_dat/timescales_bad_weights.dat')
Expected:
    Traceback (most recent call last):
        ...
    ValueError: BuildCheck 'validate_time_weights' identified error with index '2020'
Got:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest timescales.define_components[3]>", line 1, in <module>
        instance = model.create('test_dat/timescales_bad_weights.dat')
    NameError: name 'model' is not defined
Trying:
    from switch_mod.utilities import define_AbstractModel
Expecting nothing
**********************************************************************
File "timescales.py", line 352, in timescales.load_inputs
Failed example:
    from switch_mod.utilities import define_AbstractModel
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest timescales.load_inputs[0]>", line 1, in <module>
        from switch_mod.utilities import define_AbstractModel
    ImportError: No module named switch_mod.utilities
Trying:
    model = define_AbstractModel('timescales')
Expecting nothing
**********************************************************************
File "timescales.py", line 353, in timescales.load_inputs
Failed example:
    model = define_AbstractModel('timescales')
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest timescales.load_inputs[1]>", line 1, in <module>
        model = define_AbstractModel('timescales')
    NameError: name 'define_AbstractModel' is not defined
Trying:
    instance = model.load_inputs(inputs_dir='test_dat')
Expecting nothing
**********************************************************************
File "timescales.py", line 354, in timescales.load_inputs
Failed example:
    instance = model.load_inputs(inputs_dir='test_dat')
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest timescales.load_inputs[2]>", line 1, in <module>
        instance = model.load_inputs(inputs_dir='test_dat')
    NameError: name 'model' is not defined
Trying:
    instance.tp_weight_in_year.pprint()
Expecting:
    tp_weight_in_year : Size=7, Index=TIMEPOINTS, Domain=PositiveReals, Default=None, Mutable=False
        Key : Value
          1 : 1095.744
          2 : 1095.744
          3 : 1095.744
          4 : 1095.744
          5 :   2191.5
          6 :   2191.5
          7 :   8766.0
**********************************************************************
File "timescales.py", line 355, in timescales.load_inputs
Failed example:
    instance.tp_weight_in_year.pprint()
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1289, in __run
        compileflags, 1) in test.globs
      File "<doctest timescales.load_inputs[3]>", line 1, in <module>
        instance.tp_weight_in_year.pprint()
    NameError: name 'instance' is not defined
**********************************************************************
3 items had failures:
   3 of   3 in timescales
   4 of   4 in timescales.define_components
   4 of   4 in timescales.load_inputs
11 tests in 3 items.
0 passed and 11 failed.
***Test Failed*** 11 failures."
And when I run the command "python utilities_test.py" from the tests directory I get the output: 

"Traceback (most recent call last):
  File "utilities_test.py", line 6, in <module>
    import switch_mod.utilities as utilities
ImportError: No module named switch_mod.utilities".

**Comment by josiahjohnston** [comment]
That indicates that python is not picking up the switch_mod package directory. Usually this is covered by setting PYTHONPATH to the directory that the package is in. On my machine, it looks like this:
```bash
siah-macbookpro:switch_mod josiah$ pwd
/Users/josiah/src/switch/switch_mod
siah-macbookpro:switch_mod josiah$ echo $PYTHONPATH
:/Users/josiah/src/switch
siah-macbookpro:switch_mod josiah$ python -m doctest timescales.py
siah-macbookpro:switch_mod josiah$
```
Or alternately,
```bash
siah-macbookpro:switch_mod josiah$ python
Python 2.7.9 (default, Feb  2 2015, 18:27:36) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.56)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import switch_mod.utilities
>>> 
```

If you set PYTHONPATH on the command line inside a terminal window, it is only set within that context and won't retain its value if you close that terminal or open a new window. To make it persistent, you have to copy it to your login script. The only reason I can think that you got this output is that you used a different terminal window from yesterday, so PYTHONPATH became unset. If PYTHONPATH is set correctly and python still can't find the modules, then you'll need to do some troubleshooting.

If you end up having to troubleshoot the basic search path issues, here are some readings that could be useful:
* https://users-cs.au.dk/chili/PBI/pythonpath.html
* https://docs.python.org/2/tutorial/modules.html#the-module-search-path
* http://stackoverflow.com/questions/19917492/how-to-use-pythonpath
* https://docs.python.org/2/using/cmdline.html#environment-variables

Getting the search path working is an important part of installatino, but it doesn't explain why run_tests.py doesn't work. Even when I unset PYTHONPATH on my machine, run_test.py still works for me.

```bash
siah-macbookpro:switch josiah$ unset PYTHONPATH
siah-macbookpro:switch josiah$ echo $PYTHONPATH

siah-macbookpro:switch josiah$ ./run_tests.py 
................................
----------------------------------------------------------------------
Ran 32 tests in 1.149s

OK
siah-macbookpro:switch josiah$  export PYTHONPATH=/Users/josiah/src/switch
siah-macbookpro:switch josiah$ ./run_tests.py 
................................
----------------------------------------------------------------------
Ran 32 tests in 1.164s

OK
siah-macbookpro:switch josiah$
```

**Comment by mseaborn** [comment]
My guess is that the way run_tests.py overrides the discover() method isn't working on Python 2.7.3, and that's why it doesn't find the tests.  I will need to check that, though.

**Comment by rhaubenstock** [comment]
Thank you Josiah for the clarification on that PYTHONPATH if one just sets the path in terminal it will get unset when the terminal window is closed. 

I also tried to access the .profile login script, but it didn't work. Apparently instead of a .profile login script I have a .bash_profile login script, which overrides the .profile login script as mentioned here: http://stackoverflow.com/questions/6751252/difference-between-profile-and-bash-profile-on-snow-leopard. 

Furthermore, when I entered the line " export PYTHONPATH="${PYTHONPATH}:/Users/reedHaubenstock/Desktop/switch_py/" " 
into the .bash_profile file, running the command "python -m doctest -v timescales.py" in the switch_mod directory resulted in errors. However when I removed the quotations, changing the line to " export PYTHONPATH=${PYTHONPATH}:/Users/reedHaubenstock/Desktop/switch_py/ ", 
the command "python -m doctest -v timescales.py" passed all of the tests. 

However, as you said earlier, there seems to be no connection to the run_tests.py problem as it hasn't been solved for me yet. I will look into learning about the discover() method and Python 2.7.3 in the meantime.
